File: 20180712/Degree of an Array.java
class Solution {
    public int findShortestSubArray(int[] nums) {
        HashMap<Integer, ArrayList<Integer>> dic = new HashMap<Integer, ArrayList<Integer>>();
        for(int i = 0; i < nums.length; i++){
            if (dic.containsKey(nums[i])){
                dic.get(nums[i]).add(i);
            } else {
                dic.put(nums[i], new ArrayList<Integer>());
                dic.get(nums[i]).add(i);
            }
        }
        int frequency = 0;
        int res = nums.length;
        for(Map.Entry<Integer, ArrayList<Integer>> entry: dic.entrySet()) {
            ArrayList<Integer> value = entry.getValue();
            if (value.size() > frequency){
                frequency = value.size();
                res = value.get(value.size() -1 ) - value.get(0) + 1;
            }
            else if (value.size() == frequency){
                res = Math.min(res, value.get(value.size() -1 ) - value.get(0) + 1);
            }
        }
        return res;   
    }
}

File: 20180712/Degree of an Array.py
class Solution(object):
    def findShortestSubArray(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        dic = {}
        for i in range(len(nums)):
            if nums[i] in dic:
                dic[nums[i]].append(i)
            else:
                dic[nums[i]] = [i]
        frequency = 0
        res = len(nums)
        for key, value in dic.items():
            if (len(value) > frequency):
                frequency = len(value)
                res = value[-1] - value [0] + 1
            elif( len(value) == frequency):
                res = min(res, value[-1] - value [0] + 1)
        return res
        